platform/android/src/style/sources/custom_geometry_source.cpp

```diff
@@ -0,0 +1,122 @@
+#include "custom_geometry_source.hpp"
+
+#include <mbgl/renderer/query.hpp>
+
+// Java -> C++ conversion
+#include "../android_conversion.hpp"
+#include "../conversion/filter.hpp"
+//#include "../conversion/geojson.hpp"
+
+// C++ -> Java conversion
+#include "../../conversion/conversion.hpp"
+#include "../../conversion/collection.hpp"
+#include "../../geojson/conversion/feature.hpp"
+#include <mbgl/style/conversion/custom_geometry_source_options.hpp>
+
+#include <string>
+
+namespace mbgl {
+namespace android {
+
+    // This conversion is expected not to fail because it's used only in contexts where
+    // the value was originally a GeoJsonOptions object on the Java side. If it fails
+    // to convert, it's a bug in our serialization or Java-side static typing.
+    static style::CustomGeometrySource::Options convertCustomGeometrySourceOptions(jni::JNIEnv& env, jni::Object<> options, style::TileFunction fetchFn, style::TileFunction cancelFn) {
+        using namespace mbgl::style::conversion;
+        if (!options) {
+            return style::CustomGeometrySource::Options();
+        }
+        Error error;
+        optional<style::CustomGeometrySource::Options> result = convert<style::CustomGeometrySource::Options>(Value(env, options), error);
+        if (!result) {
+            throw std::logic_error(error.message);
+        }
+        result->fetchTileFunction = fetchFn;
+        result->cancelTileFunction = cancelFn;
+        return *result;
+    }
+
+CustomGeometrySource::CustomGeometrySource(jni::JNIEnv& env, jni::Object<CustomGeometrySource> _obj, jni::String sourceId, jni::Object<> options)
+    : Source(env, std::make_unique<mbgl::style::CustomGeometrySource>(
+            jni::Make<std::string>(env, sourceId),
+            convertCustomGeometrySourceOptions(env,
+                                               options,
+                                               std::bind(&CustomGeometrySource::fetchTile, this, std::placeholders::_1),
+                                               std::bind(&CustomGeometrySource::cancelTile, this, std::placeholders::_1))) ),
+      javaPeer(_obj.NewGlobalRef(env))  {
```

Comment:
I don't think this is going to work properly as the Java/c++ peers will never be cleaned up.

As far as I can tell, the java object is created, which creates the c++ peer. The c++ peer holds a global reference to the java peer, ensuring it is not GC'd. Now, the only way to clean up both would be to destroy the c++ peer, which would release the global reference on the Java peer, making it eligible for GC. This in turn would however try to call the destructor of the c++ peer as that is how it has been registered through `jni::RegisterNativePeer` below.

For this to work, the life-cycle of the the c++ peer should be coupled to the core `CustomGeometrySource`. When it is destroyed, the c++ peer should be destroyed, releasing the global reference, making the java peer eligible for GC.